### DISCUSSION TITLE
tar_make speed

### DISCUSSION THREAD

**Discussion by llrs** [ropensci/targets#914] [General]
I am using targets quite interactively, and I was surprised to note that `tar_make()` takes a "long time" to finish even when nothing is done and everything is skiped.
Making the same target 100 times, resulted in around 4 seconds to finish  but there was nothing to do (in interactive sessions I think the median time is almost half, but that might be a completely different issue). I understand that this might depend on several things (complexity of the pipe probably, how I use the package...) but maybe it could be faster.

```r
# A tibble: 1 × 13
  expression                                min median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time result
  <bch:expr>                              <bch> <bch:>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm> <list>
1 tar_make(cAscending_vs_Sigma_colon_ASV) 3.86s  3.97s     0.250    1.98MB   0.0587    81    19       5.4m <NULL>
```

Many thanks, again for the wonderful package Will.

**Comment by wlandau** [comment]
It's tough to say a priori, but common bottlenecks are a slow file system (e.g. `_targets/` on a Windows network drive), a large `renv` library that is slow to activate, overhead due to `callr::r()` creating a reproducible R process for the pipeline, a large number of targets, a large number of `tar_quarto()` calls (https://github.com/quarto-dev/quarto-cli/discussions/1266), and other stuff that happens when the pipeline is setting up. To find out empirically, you could profile your pipeline: https://books.ropensci.org/targets/performance.html#profiling

**Reply by llrs** [reply]
I am on Ubuntu 20.04, renv library has 314 packages (57 direct dependencies in _target_packages.R), the target doesn't have any quarto calls (R 4.1.2). However, I couldn't profile the process via `proffer::pprof` as I got an error:

```
Error in rethrow_call(c_processx_exec, command, c(command, args), pty,  : 
  cannot start processx process '' (system error 2, No such file or directory) @unix/processx.c:608 (processx_exec)
```

Using Rstudio profiler I got the following:
![Screenshot from 2022-08-30 10-33-25](https://user-images.githubusercontent.com/6818218/187393572-8aeeddae-aa71-402f-bdb3-c05dc2d987b2.png)

Despite the CPU(?) time of just 180 ms the output said `✓ skip pipeline: 1.472 seconds`.
I could share the code but without the data ,which I cannot share, it wouldn't be worth much. Anyway, this is a minor inconvenience and probably it is not worth dedicating more time.

**Reply by wlandau** [reply]
Looks like you shaved off a couple seconds with `callr_function = NULL`. Other overhead might be due to loading packages. If you put them in `tar_option_set(packages = c("..."))` instead of calling `library(pkg)` in `_targets.R`, that could save time if the targets that use those packages are skipped.

That `unix/processx.c` error is strange, maybe you would need to upgrade the `processx` package?

**Reply by llrs** [reply]
I already call each dependency on each target via `tar_target(..., packages=c("..."))`. I'll time what I have in _targets.R. Many thanks as always!  
I'll check what's happening with the processx package.
